clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonApplicationLoadBalancerCachingAgent.groovy

```diff
@@ -298,12 +473,75 @@ class AmazonApplicationLoadBalancerCachingAgent extends AbstractAmazonLoadBalanc
         }
       }
     }
+
+    List<InstanceTargetGroupState> itgStates = []
+    for (TargetGroup tg : allTargetGroups) {
+      // Collect health information for the target group and instance ids
+      List<String> instanceIds = new ArrayList<String>()
+      List<TargetHealthDescription> thds = targetGroupArnToHealths.get(tg.targetGroupArn)
+      for (TargetHealthDescription thd : thds) {
+        itgStates << new InstanceTargetGroupState(thd.target.id, ArnUtils.extractTargetGroupName(tg.targetGroupArn).get(), thd.targetHealth.state, thd.targetHealth.reason, thd.targetHealth.description)
+        instanceIds << thd.target.id
+      }
+
+      // Get associated load balancer keys
+      Collection<String> loadBalancerIds = tg.loadBalancerArns.collect {
+        String lbName = ArnUtils.extractLoadBalancerName(it).get()
+        String lbType = ArnUtils.extractLoadBalancerType(it)
+        Keys.getLoadBalancerKey(lbName, account.name, region, tg.vpcId, lbType)
+      }
+
+      // Get target group attributes
+      Map<String, String> tgAttributes = targetGroupArnToAttributes.get(tg.targetGroupArn).collectEntries {
+        [(it.key): it.value]
+      }
+
+      Map<String, Object> tgCacheAttributes = objectMapper.convertValue(tg, ATTRIBUTES)
+      tgCacheAttributes.put('instances', instanceIds)
+      tgCacheAttributes.put('attributes', tgAttributes)
+      tgCacheAttributes.loadBalancerNames = tgCacheAttributes.loadBalancerArns.collect { String lbArn ->
+        ArnUtils.extractLoadBalancerName(lbArn).get()
+      }
+      tgCacheAttributes.remove('loadBalancerArns')
+
+      // Cache target group
+      String targetGroupId = Keys.getTargetGroupKey(tg.targetGroupName, account.name, region, tg.vpcId)
+      targetGroups[targetGroupId].with {
+        attributes.putAll(tgCacheAttributes)
+        relationships[LOAD_BALANCERS.ns].addAll(loadBalancerIds)
+      }
+      for (String loadBalancerId : loadBalancerIds) {
+        loadBalancers[loadBalancerId].with {
+          relationships[TARGET_GROUPS.ns].add(targetGroupId)
+        }
+      }
+    }
+
+    // Build health cache
+    // Have to do this separately because an instance can be in multiple target groups
+    List<InstanceTargetGroups> itgs = InstanceTargetGroups.fromInstanceTargetGroupStates(itgStates)
+    Collection<CacheData> tgHealths = []
+    Collection<CacheData> instances = []
+
+    for (InstanceTargetGroups itg in itgs) {
+      String instanceId = Keys.getInstanceKey(itg.instanceId, account.name, region)
+      String healthId = Keys.getInstanceHealthKey(itg.instanceId, account.name, region, healthId)
```

Comment:
Have that class level attribute called `healthId` is pretty shitty here because you've now hidden it.